tests/unit_tests/test_type_deductor.py

```diff
@@ -0,0 +1,151 @@
+        }, index=list(range(n_points)))
+
+        input_data = TransactionData()
+        input_data.data_frame = input_dataframe
+        type_deductor.run(input_data)
+
+        stats_v2 = lmd['stats_v2']
+
+        assert stats_v2['numeric_id']['is_foreign_key']
+        assert stats_v2['uuid']['is_foreign_key']
+
+        assert 'numeric_id' in lmd['columns_to_ignore']
+        assert 'uuid' in lmd['columns_to_ignore']
+
+    def test_empty_column(self, transaction, lmd):
+        type_deductor = TypeDeductor(session=transaction.session,
+                                     transaction=transaction)
+
+        n_points = 100
+        input_dataframe = pd.DataFrame({
+            'numeric_int': list(range(n_points)),
+            'empty_column': [None for i in range(n_points)],
+        }, index=list(range(n_points)))
+
+        input_data = TransactionData()
+        input_data.data_frame = input_dataframe
+        type_deductor.run(input_data)
+
+        stats_v2 = lmd['stats_v2']
+        assert stats_v2['empty_column']['typing']['data_type'] is None
+
+    def test_empty_values(self, transaction, lmd):
+        type_deductor = TypeDeductor(session=transaction.session,
+                                    transaction=transaction)
+
+        n_points = 100
+        input_dataframe = pd.DataFrame({
+            'numeric_float': np.linspace(0, n_points, n_points),
+        }, index=list(range(n_points)))
+        input_dataframe['numeric_float'].iloc[::2] = None
+        input_data = TransactionData()
+        input_data.data_frame = input_dataframe
+        type_deductor.run(input_data)
+
+        stats_v2 = lmd['stats_v2']
+        assert stats_v2['numeric_float']['typing']['data_type'] == DATA_TYPES.NUMERIC
+        assert stats_v2['numeric_float']['typing']['data_subtype'] == DATA_SUBTYPES.FLOAT
+        assert stats_v2['numeric_float']['typing']['data_type_dist'][DATA_TYPES.NUMERIC] == 50
+        assert stats_v2['numeric_float']['typing']['data_subtype_dist'][DATA_SUBTYPES.FLOAT] == 50
+
+    def test_type_mix(self, transaction, lmd):
+        type_deductor = TypeDeductor(session=transaction.session,
+                                     transaction=transaction)
+
+        n_points = 100
+        input_dataframe = pd.DataFrame({
+            'numeric_float': np.linspace(0, n_points, n_points),
+        }, index=list(range(n_points)))
+        input_dataframe['numeric_float'].iloc[:2] = 'random string'
+        input_data = TransactionData()
+        input_data.data_frame = input_dataframe
+        type_deductor.run(input_data)
+
+        stats_v2 = lmd['stats_v2']
+        assert stats_v2['numeric_float']['typing']['data_type'] == DATA_TYPES.NUMERIC
+        assert stats_v2['numeric_float']['typing']['data_subtype'] == DATA_SUBTYPES.FLOAT
+        assert stats_v2['numeric_float']['typing']['data_type_dist'][DATA_TYPES.NUMERIC] == 97
+        assert stats_v2['numeric_float']['typing']['data_subtype_dist'][DATA_SUBTYPES.FLOAT] == 97
```

Comment:
I'm getting some warning about `no newline at the end of file` here, but I'm not sure this is relevant ?

Reply:
Ah yes, empty line at the end of file is a convention I always forget to follow :) fixed